src/qbe_ir_generator.cpp

```diff
@@ -126,11 +126,57 @@ void QbeIrGenerator::Visit(const DeclNode& decl) {
   id_to_num[decl.id] = id_num;
 }
 
+void QbeIrGenerator::Visit(const ParameterNode& parameter) {
+  int id_num = NextLocalNum();
+  id_to_num[parameter.id] = id_num;
+  int reg_num = NextLocalNum();
+  num_recorder.Record(reg_num);
+}
+
 void QbeIrGenerator::Visit(const FuncDefNode& func_def) {
   int label_num = NextLabelNum();
+  // parameters initilization before going into the body
   auto start_label = BlockLabel{"start", label_num};
-  Write_("export function w ${}() {{\n", func_def.id);
-  WriteLabel_(start_label);
+  auto body_label = BlockLabel{"body", label_num};
+
+  Write_("export\n");
+  if (func_def.parameters.empty()) {
+    Write_("function w ${}() {{\n", func_def.id);
+    WriteLabel_(start_label);
+  } else {
+    // @note: In order to produce the correct function signature,
+    // e.g., function w $add(w %.1, w %.3, w %.5), we need to visit every
+    // parameter and store their id_num and reg_num, so that we can generate
+    // the correct instructions after the function signature.
+    std::string func_params;
+    std::map<std::string, int> id_to_reg_num;
+    for (const auto& parameter : func_def.parameters) {
+      parameter->Accept(*this);
+      int reg_num = num_recorder.NumOfPrevExpr();
+      id_to_reg_num[parameter->id] = reg_num;
+      // TODO: support different data types
+      func_params.append(fmt::format("w %.{}", id_to_num[parameter->id]));
```

Comment:
```suggestion
      func_params.append(fmt::format("w %.{}", id_to_num.at(parameter->id)));
```

Please consider using the `at` function for all element access. It provides bound checking, throwing an error instead of causing a segmentation fault when the element does not exist.